### PR TITLE
private sector datamigration

### PR DESCRIPTION
**Pull request by nickpell**
This is a work in progress, but figured it would be good to merge with master and get feedback.

The management command ```create_cases_from_beneficiary``` is run to populate person, occurrence, episode, and adherence cases based off of the data dumps that we get from UATBC (imported into postgres).  Prescription and test cases are included for tests but I'm not populating those case properties yet, since they have low priority.

@proteusvacuum @mkangia cc: @orangejenny

### CODE REVIEW

File: custom/enikshay/private_sector_datamigration/models.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        return {
+            'Positive': 'reactive',
+            'Negative': 'non_reactive',
+            'Not known': 'unknown',
+            'Select': 'unknown',
+            None: 'unknown',
+        }[self.hiv]
+
+    @property
+    def site_choice(self):
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/models.py]: where did you get this list? We are using the list here in the app: https://india.commcarehq.org/a/enikshay/fixtures/view_lookup_tables/?table_id=790cfc46c50b4f41bb090c1f3c5b8dad

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        return episode_or_descendants + tests
+
+    def get_person_case_structure(self):
+        kwargs = {
+            'attrs': {
+                'case_type': PERSON_CASE_TYPE,
+                'close': False,
+                'create': True,
+                'update': {
+                    'current_address': self.beneficiary.current_address,
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: can you add `enrolled_in_private = 'true'`

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+                }
+            }
+        }
+
+        if self.beneficiary.age_entered is not None:
+            kwargs['attrs']['update']['age'] = self.beneficiary.age_entered
+            kwargs['attrs']['update']['age_entered'] = self.beneficiary.age_entered
+        else:
+            if self.beneficiary.dob is not None:
+                kwargs['attrs']['update']['age'] = relativedelta(
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: nice

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            kwargs['attrs']['update']['sex'] = self.beneficiary.sex
+
+        if self.beneficiary.has_aadhaar_number:
+            kwargs['attrs']['update']['aadhaar_number'] = self.beneficiary.identificationNumber
+        else:
+            kwargs['attrs']['update']['other_id_number'] = self.beneficiary.identificationNumber
+            kwargs['attrs']['update']['other_id_type'] = self.beneficiary.other_id_type
+
+        if self._episode:
+            kwargs['attrs']['update']['hiv_status'] = self._episode.hiv_status
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: do they have `diabetes_status` too?

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+    def get_episode_case_structure(self, occurrence_structure):
+        kwargs = {
+            'attrs': {
+                'case_type': EPISODE_CASE_TYPE,
+                'close': False,
+                'create': True,
+                'owner_id': '-',
+                'update': {
+                    'adherence_schedule_id': 'schedule_mwf',
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: Need to check if this is true (I'm not sure)

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+                    'date_of_mo_signature': self.beneficiary.dateOfRegn.date(),
+                    'dots_99_enabled': 'false',
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: I think there might be some patients already enrolled in 99DOTS in their system... Need to think how we're going to link those up.

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                    'episode_id': get_human_friendly_id(),
+                    'episode_type': self.beneficiary.current_episode_type,
+                    'name': self.beneficiary.episode_name,
+                    'transfer_in': '',
+
+                    'migration_created_case': 'true',
+                    'migration_created_from_record': self.beneficiary.caseId,
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: can you add `enrolled_in_private = true`

File: custom/enikshay/private_sector_datamigration/factory.py [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+                'update': {
+                    'name': prescription.productName,
+
+                    'migration_created_case': 'true',
+                    'migration_created_from_record': prescription.prescriptionID,
+                }
+            },
+            'indices': [CaseIndex(
+                episode_structure,
+                identifier='host',
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/factory.py]: prescriptions use `episode_of_prescription` as the identifier

File: custom/enikshay/private_sector_datamigration/management/commands/create_cases_by_beneficiary.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        parser.add_argument(
+            '--nikshayId',
+            dest='nikshayId',
+            default=None,
+        )
+
+    @mock_ownership_cleanliness_checks()
+    def handle(self, domain, **options):
+        base_query = Beneficiary.objects.filter(
+            caseStatus__in=['suspect', 'patient', 'patient '],
```
Comment (proteusvacuum) [inline custom/enikshay/private_sector_datamigration/management/commands/create_cases_by_beneficiary.py]: ouch.

**Review by proteusvacuum — APPROVED** [review]
Yo, this looks awesome.

Have we done an exercise to check that this populates all the case properties that are collected by the app if possible? 
I think you should be able to use: https://india.commcarehq.org/a/enikshay/apps/view/9340429733463e58ae0e1518defee221/summary/#/cases

I can help with that if you want. I think we should be explicit in the items that we are leaving out from the migration / or aren't collecting in the app that are present in the UATBC data.

**Comment by proteusvacuum** [comment]
Merging for now as we're going to start gathering conflicts in our test files and such, and this shouldn't affect anything else.

File: custom/enikshay/private_sector_datamigration/models.py (hunk 1 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/models.py]: For the keys, I checked what values were in the data dump.

For the values, referenced ```site_choice``` from the public sector: https://india.commcarehq.org/a/enikshay/apps/view/ce3e9160e05acd8e0020ed5e33ca89ea/summary/#/cases

It looks like each list has values missing in the other one - is that expected?

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 2 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: https://github.com/dimagi/commcare-hq/pull/16015

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 4 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: https://github.com/dimagi/commcare-hq/pull/16015

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 5 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: added todo https://github.com/dimagi/commcare-hq/pull/16015

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 6 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: added todo https://github.com/dimagi/commcare-hq/pull/16015

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 7 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: https://github.com/dimagi/commcare-hq/pull/16015

File: custom/enikshay/private_sector_datamigration/factory.py (hunk 8 above)
Comment (nickpell) [inline custom/enikshay/private_sector_datamigration/factory.py]: https://github.com/dimagi/commcare-hq/pull/16015

**Pull request merged**